### ISSUE TITLE
Component/declarative error handling (ala React 16's <ErrorBoundary/>)

### ISSUE THREAD

**Issue by craigsmitham** [angular/angular#18509]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

React recently introduced a concept called "Error Boundaries" (https://facebook.github.io/react/blog/2017/07/26/error-handling-in-react-16.html). Basically you wrap key areas of your app with <ErrorBoundary> .... </ErrorBoundary> and any component/template errors within can be handled at that level. I've run into issues with my Angular 4 app where an error in a component will cause the app to break ... sometimes the app is still navigable but often times rendering is messed up and the app just needs a restart. Most blog posts (and the Angular documentation) on error handling don't deal with rendering issues, just logging/handling any exceptions that may occur. What are strategies for gracefully handling "breaking errors" in an Angular app? Would something like <ErrorBoundary/> make sense for Angular?

**Comment by mtuzinskiy** [comment]
This would be very helpful. The fact that in Angular entire app "sinks" on a single error in some component is just embarrassing :-1:

**Comment by mtuzinskiy** [comment]
Any updates on this?

**Comment by pauldraper** [comment]
**error-boundary.component.ts**

```typescript
import {
  ChangeDetectorRef,
  Component,
  ErrorHandler,
  ContentChild,
  TemplateRef,
} from '@angular/core';

@Component({
  selector: 'error-boundary',
  templateUrl: './error-boundary.component.html',
})
export class ErrorBoundaryComponent {
  constructor(
    private readonly changeDetector: ChangeDetectorRef,
    private readonly errorHandler: ErrorHandler,
  ) {
    this.changeDetector.detach();
  }

  @ContentChild(TemplateRef)
  template: TemplateRef<{}>;

  ngDoCheck() {
    try {
      this.changeDetector.detectChanges();
    } catch (e) {
      this.errorHandler.handleError(e);
    }
  }
}
```

**error-boundary.component.html**

```html
<ng-container *ngTemplateOutlet="template"></ng-container>
```

**Comment by sandrooco** [comment]
@pauldraper I'm trying to implement this with a child component from a libray (into which I obviously can't build error handling). Do you have an idea how I could achieve this?

**Comment by mtuzinskiy** [comment]
@sandrooco  obviously, there is no way. This is not a problem that can be easily solved by some component. It's a global serious  problem of Angular framework, which is being consistently ignored by the Angular Team.

**Comment by shivekundefined** [comment]
There is one npm package that can help us to achieve error handling on component just like React does.
https://www.npmjs.com/package/ngx-error-boundary
This should be helpful

**Comment by mtuzinskiy** [comment]
@shivekundefined  No, that's a totally different thing. I checked the code, and this library basically just provides a service with a "handleError" method + a component to do something with an error. @pauldraper 's answer above is much shorter and useful.  
**BUT**, we're not talking about mere error handling here. We're talking about low-level Angular functionality: ability to ignore the errors while rendering the components, and continue to run the app gracefully.  (remember that you have 3rd party components from OTHER libs, too).  
That's why it says **Error Boundaries**: the error has power only in limited boundaries. We're not saying "Error Handling" :slightly_smiling_face:

**Comment by crutchcorn** [comment]
I believe I have an API that would solve this issue:

https://github.com/angular/angular/issues/51941

Would love to hear your thoughts/thumbs up/down on the matter. Even wrote a blog post (linked in issue) why it's needed

**Comment by bredziniak** [comment]
This would be great, it could also be achieved by new control syntax, something like: 
@try {
<component-that-throw/>
} @catch(e) {
<error-card [error]="e" />
}

**Comment by hv0905** [comment]
Hi, I’m looking forward to seeing this feature. My application has several components that process user-provided data, and handling errors within them is complex. I’m hoping for a way to manage errors outside of the components.

**Comment by menosprezzi** [comment]
I was thinking how it will blend with the `@defer` block syntax since it has the `@error` block. I think it will be misunderstood. I've always been in favor that `@error` should be called `@catch`, even though it cant catch errors throwed by the component, but aiming to support features like this as soon as possible, without generating confusion with a functionality like this, converging on a single taxonomy.

If it does not converge, the markup of a deferrable view with propper error handling will be something like this 😵‍💫:
```html
@defer {
    @try {
	<error-throwing />
    } @catch {
	<handle-error [error]="$error.message"/>
    } 
}
@loading {
    <loading />
}
@error {
    <handle-error [error]="$error.message"/>
}
```

**Comment by raven-dever** [comment]
+1 on this. I think it's a quite needed feature to ensure the stability of Angular apps

**Comment by SSE-programmer** [comment]
I agree. We need a solution to this.

**Comment by chriswoodie** [comment]
The new @boundary/@error syntax solves this, but will it work for images whose `src` fails to load to allow us to provide an error image instead?
